Initial implementation

alvarosaraujo opened the pull request.

silverwings88 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Hyro training exercise: Dynamic Properties](https://trello.com/c/RpnlKfYb/7-hyro-training-exercise-dynamic-properties)

silverwings88 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Hyro training exercise: Dynamic Properties](https://trello.com/c/RpnlKfYb/7-hyro-training-exercise-dynamic-properties)

Commit:
Created new repository

Commit:
Dynamic Properties test completed

joaobrittoneto commented on CMakeLists.txt:
```diff
@@ -0,0 +1,70 @@
+hyro_init_package(hyro_signal_generator VERSION 0.0.1)
+
+include ( FindPkgConfig  )
+find_package ( Qt4 COMPONENTS QtCore QtGui QtOpenGl )
```
Do you use Qt? If not you can remove Qt related code from here.

joaobrittoneto commented on package.json:
```diff
@@ -0,0 +1,13 @@
+	"description": "Signal generator and processor",
+
+	"dependencies": {
+		"hyro_api_channel": "*",
```
Missing dependencies.
Usually all packages required in the CMakeList.txt need to be in the dependency list of package.json

joaobrittoneto commented on include/hyro/DigitalConverterComponent.h:
```diff
@@ -0,0 +1,75 @@
+
+private:
+
+  double amplitude, threshold;
```
Usually we use `m_variable` to indicate a member of the class

joaobrittoneto commented on include/hyro/msgs/Signal.h:
```diff
@@ -0,0 +1,23 @@
+
+struct HYRO_SIGNAL_GENERATOR_COMPONENTS_EXPORT Signal
+{
+  std::string timestamp;
```
Don't represent timestamp as string.

 Use `uint64`, or `hyro::Time`

joaobrittoneto commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+bool 
+DigitalConverterComponent::setThreshold(double th) 
+{ 
+  threshold = th; 
```
There is no need to have a `threshold` here. We can use just the one in `thold`

joaobrittoneto commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+double 
+DigitalConverterComponent::getThreshold() 
+{ 
+  return threshold; 
```
try to do `return thold.getThreshold();`

joaobrittoneto commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+bool 
+DigitalConverterComponent::setAmplitude(double amp) 
+{
+  amplitude = amp;
```
Same as above

joaobrittoneto commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+double 
+DigitalConverterComponent::getAmplitude() 
+{ 
+  return amplitude; 
```
Same as above

joaobrittoneto commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+DigitalConverterComponent::callback (std::shared_ptr<const Signal> && value)
+{
+  //Retrieves data from received message
+  std::string timestamp = value->timestamp;
```
It seems you don't use timestamp and frame here. If it's the case, just remove them from here.

joaobrittoneto commented on src/internal/Thresholding.cpp:
```diff
@@ -0,0 +1,31 @@
+}
+
+
+double Thresholding::get_threshold(double value)
```
Be consistent with the names. Use `getThreshold`

joaobrittoneto commented on src/internal/Thresholding.cpp:
```diff
@@ -0,0 +1,31 @@
+double Thresholding::get_threshold(double value)
+{
+  if (value > threshold)
+  { return amplitude; }
```
There is no need of `{ }` for a single line.

```
if (value > threshold)
  return amplitude;
return 0;
```

joaobrittoneto commented on include/hyro/Thresholding.h:
```diff
@@ -0,0 +1,28 @@
+{
+  private:
+
+    double amplitude, threshold;
```
Usually the member of class are named with a suffix `m_`

In this case, `m_amplitude` and `m_threshold`

gustavoneves12 commented on include/hyro/SignalGenerator.h:
```diff
@@ -0,0 +1,33 @@
+{
+  private:
+
+    double rad, m_deg, m_step, m_amp;
```
Please, include prefix `m_` for class member variables.

gustavoneves12 commented on msgs/hyro/msgs/Signal.proto:
```diff
@@ -0,0 +1,11 @@
+{
+    double value        = 1;
+    string frame_id     = 2;
+    string timestamp    = 3;
```
Use `int64` for `timestamp`.

gustavoneves12 commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace hyro
+{
+hyro::Signal message_signal_rec;
```
Where are you using these variables? Please, avoid to maintain unused variables.

gustavoneves12 commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+{
+hyro::Signal message_signal_rec;
+hyro::Signal message_output;
+Thresholding thold;
```
This variable must be a member class variable.

Please, do not use global variables.

alvarosaraujo replied:
The variable `Thresholding thold` is used in different methods, such as the getters and setters. Should I declare this variable inside each function, every time I call one of these methods?

gustavoneves12 commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+
+Result
+DigitalConverterComponent::init (const ComponentConfiguration & configuration)
+{
```
This component has two parameters **amplitude** and **threshold**. Are you initializing these parameters?

The components' parameters must be initialized using the function ` config.parameters.getParameter`.

gustavoneves12 commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+
+Result
+SignalGeneratorComponent::init (const ComponentConfiguration & configuration)
+{
```
This component has four parameters **amplitude**, **frequency**, **name** and **cosine**.  These parameters must be initialized using the function ` config.parameters.getParameter`.

alvarosaraujo replied:
Sorry, but I'm not exactly sure how to implement this. These variables are members of SignalGeneratorComponent class. Aren't they accessible when they're declared in the .h file?

gustavoneves12 replied:
According to the exercise the **SignalGeneratorComponent** has four parameters (amplitude, frequency, name and cosine).

The parameters are define in the component connection string. Because of that, you need to call the function `config.parameter.getParameter` to load the parameters values.

gustavoneves12 commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+                                this
+                                );
+
+  std::shared_ptr<ChannelOutput<std::vector<int>>>
```
you can replace this to `auto`.

```c++
auto m_dummy = registerOutput<std::vector<int>>("fix_dynamic"_uri, configuration);
```

gustavoneves12 commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+                                );
+
+  std::shared_ptr<ChannelOutput<std::vector<int>>>
+  m_dummy = this->registerOutput<std::vector<int>>("fix_dynamic"_uri, configuration);
```
You don't need to use `this` all the time. I usually just use `this` when it is required.

gustavoneves12 commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+double
+SignalGeneratorComponent::getAmplitude()
+{
+  return amplitude;
```
As mentioned by @joaobrittoneto you can use `sg.getAmplitude()`. Avoid to declare redundant variables and take advantages of the encapsulation provided by the classes.

gustavoneves12 commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+double
+SignalGeneratorComponent::getFrequency()
+{
+  return frequency;
```
The same here.

gustavoneves12 commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+  hyro::Signal message_signal;
+  time_t t = std::time(0);
+
+  message_signal.timestamp = std::asctime(std::localtime(&t));
```
Do not use string for `timestamp`. Use `uint64` or `hyro::Time`.

gustavoneves12 commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,119 @@
+}
+
+Result
+DigitalConverterComponent::check ()
```
Please check if the connection was established with success here.

```c++
  if (m_input->connectionStatus() != ConnectionStatus::CONNECTION_UP)
    return Result::RESULT_ERROR;
```

alvarosaraujo replied:
I implemented this check, it works just fine in the main code. Though it generates a failure when used in the test. Is there something else that has to be done in order to make it work in both scenarios?

```
[2019-01-31 13:57:50.240429][trace][StateMachine] [/digital] Checking -> Fault
/home/alvaro/dev/hyro/workspace/src/hyro_signal_generator/tests/test_signal_generator.cpp:51: Failure
Expected equality of these values:
  Result::RESULT_OK
    Which is: 4-byte object <00-00 00-00>
  digital_converter_state_machine.check()
    Which is: 4-byte object <02-00 00-00>
```

gustavoneves12 replied:
Your testing is failing because you need first start the output channel that will be connected before call check function.

Probably, in the test, the channel which connects to the input channel of digital_converter was not started.

gustavoneves12 commented on src/examples/signal_generator.cpp:
```diff
@@ -0,0 +1,239 @@
+  //-------------------------------------------------
+
+  // GRPC
+  #if defined(TEST_GRPC)
```
I see you have good intentions defining all the protocols here.

:+1:

gustavoneves12 commented on src/examples/signal_generator.cpp:
```diff
@@ -0,0 +1,239 @@
+using namespace std::string_literals;
+using namespace std::chrono_literals;
+
+double signal_property_amplitude;
```
@silverwings88 please, avoid global variables.

Please, try always use local variables and arguments.

gustavoneves12 commented on src/examples/signal_generator.cpp:
```diff
@@ -0,0 +1,239 @@
+  // Step 11: Start plot function as thread
+  //-------------------------------------------------
+
+  std::thread t1(plot);
```
:+1:

gustavoneves12 commented on src/examples/signal_generator.cpp:
```diff
@@ -0,0 +1,239 @@
+  //          new dynamic properties
+  //-------------------------------------------------
+
+  while (true)
```
You program will never end, just pressing`Ctrl+C` on the terminal.

I suggest you add a question **"Continue? (Y/N)"** after setting the dynamic parameters. If the user inform **"N"** stop the loop.

Other option is call the **plot()** in the main thread and move this while to a thread. In this way, you can exit the program when the user close all windows.

gustavoneves12 commented on src/internal/SignalGenerator.cpp:
```diff
@@ -0,0 +1,36 @@
+namespace hyro 
+{
+
+SignalGenerator::SignalGenerator()
```
We usually the member variables are initialized as shown below:

```c++
 SignalGenerator::SignalGenerator()
  : m_deg(0.0)
  , m_step(1.0)
  , m_amp(1.0)
{
}
```

gustavoneves12 commented on src/internal/SignalGenerator.cpp:
```diff
@@ -0,0 +1,36 @@
+  return  m_amp * sin(rad);
+}
+
+void SignalGenerator::setAmplitude(double amplitude) { m_amp = amplitude; }
```
Please, do not use this style for functions in the CPP files. You can move the getters and setters functions to the header, and the you can declare the functions using one line.

In header file you can do:

```c++
// In the .h file
class SignalGenerator
{
 void setAmplitude(double amplitude) { m_amp = amplitude; }
};
```

However, in the CPP or implementation file, you never declare a function using one line.

```c++
// In the .cpp file
void SignalGenerator::setAmplitude(double amplitude)
{ 
  m_amp = amplitude;
}
```


void SignalGenerator::setAmplitude(double amplitude) { m_amp = amplitude; }

gustavoneves12 commented on src/internal/SignalGenerator.cpp:
```diff
@@ -0,0 +1,36 @@
+
+double SignalGenerator::getAmplitude() { return m_amp; }
+
+void SignalGenerator::setFrequency(double frequency) { m_step = ( 360 * frequency * 10 ) / 1000 ; }
```
I suggest you add the decimals (360.0) when using double types. For float types use decimals followed by f (360.0f).

gustavoneves12 commented on src/internal/Thresholding.cpp:
```diff
@@ -0,0 +1,31 @@
+{
+
+Thresholding::Thresholding()
+{
```
See the code below:

```c++
Thresholding::Thresholding()
  : m_amplitude(1)
  , m_threshold(0)
{
}
```

gustavoneves12 commented on tests/test_signal_generator.cpp:
```diff
@@ -0,0 +1,216 @@
+  "}}";
+
+  //SignalGeneratorComponent and DigitalConverterComponent state machines
+  StateMachine sm(std::make_shared<SignalGeneratorComponent>("/signal"_uri));
```
Please, give meaningful name for the variables. e.g.: `signal_generator_sm`, `digital_converter_sm`.

gustavoneves12 commented on tests/test_signal_generator.cpp:
```diff
@@ -0,0 +1,216 @@
+  StateMachine sm(std::make_shared<SignalGeneratorComponent>("/signal"_uri));
+  StateMachine dc(std::make_shared<DigitalConverterComponent>("/digital"_uri));
+
+  auto result_sm = sm.init(ComponentConfiguration(signal_conf));
```
I would remove the `result_sm` variable. You can call as follows:

```c++
ASSERT_EQ(Result::RESULT_OK, sm.init(ComponentConfiguration(signal_conf));
```

gustavoneves12 commented on tests/test_signal_generator.cpp:
```diff
@@ -0,0 +1,216 @@
+//-------------------------------------------------
+// Create test for DigitalConverterComponent
+//-------------------------------------------------
+TEST (ExampleTest, DigitalConverterComponentCheck)
```
Here you need to check if the output of `DigitalConverterComponent` is corrected. Due to use a spinner to asynchronous write the digital converter output, it is difficult to check the exactly value. However, you can check if the output is between a specific range. For example, greater than 0 and less than 1.

gustavoneves12 commented on tests/test_signal_generator.cpp:
```diff
@@ -0,0 +1,216 @@
+  auto inputSignal = std::make_shared<FakeInput<Signal>>("/digital"_uri, "api", "/digital/digital");
+  ASSERT_TRUE(inputSignal->connect());
+
+  //Fake output channels
```
The fake output should be declare before call the `connect` function of **DigitalConverterComponent**.

Commit:
Qt removed
Dependecies added to package.json
Variable names corrected
Timestamp represented as uint64
Removed unused varaibles
getThreshold() constructor inside Thresholding.cpp renamed as calculateThreshold()
Removed brackets for one line functions
Check method implemented
Request input from user to exit code
Decimals added for double type variables

gustavoneves12 commented on src/internal/SignalGenerator.cpp:
```diff
@@ -0,0 +1,36 @@
+  return  m_amp * sin(m_rad);
+}
+
+// void SignalGenerator::setAmplitude(double amplitude) { m_amp = amplitude; }
```
Remove useless commentaries.

gustavoneves12 commented on tests/test_signal_generator.cpp:
```diff
@@ -0,0 +1,189 @@
+  ASSERT_EQ(Result::RESULT_OK, digital_converter_state_machine.connect(ConnectionConfiguration()));
+  //ASSERT_EQ(Result::RESULT_OK, digital_converter_state_machine.check());
+
+  // hyro::Signal message_sgn;
```
Remove useless commentaries.

joaobrittoneto commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,125 @@
+Result
+SignalGeneratorComponent::start ()
+{
+  double freq = 1;
```
This initialization goes in the `init` method, as @gustavoneves12  said. 
And avoid to hard-coded the parameters like that. In the `init` you can get the parameters from the configuration, with a default value.

```
double frequency = configuration.parameters.getParameter<float>("frequency", 1);
sg.setFrequency(frequency);
```

Commit:
Removed global variables
Removed unused variables
Moved plot function to main code and user input request to separate thread
Added parameters initialization inside init() methods
Removed useless comments
Test code fully functional

joaobrittoneto commented on src/internal/SignalGenerator.cpp:
```diff
@@ -0,0 +1,25 @@
+
+double SignalGenerator::getSignalValue()
+{
+  m_rad = m_deg / 180.0 * M_PI;
```
`m_rad` is just used here. It can be defined locally, instead of being a member of the class.

```
double rad = m_deg / 180.0 * M_PI;
```

joaobrittoneto commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,121 @@
+namespace hyro
+{
+std::shared_ptr<HyroLogger> SignalGeneratorComponent::s_logger = HyroLoggerManager::CreateLogger("SignalGeneratorComponent");
+// SignalGenerator sg;
```
Commented code can be removed

joaobrittoneto commented on include/hyro/SignalGeneratorComponent.h:
```diff
@@ -0,0 +1,79 @@
+
+  double m_amplitude, m_frequency;
+  int m_cosine;
+  SignalGenerator sg;
```
I'd suggest to give meaningful names do variables. It increases readability of the code.
You can replace `sg` by `m_signal_gen` for example.

Commit:
Removed commented code
Replaced variable names for more readability
Removed unused variables

joaobrittoneto commented on src/components/DigitalConverterComponent.cpp:
```diff
@@ -0,0 +1,115 @@
+DigitalConverterComponent::init (const ComponentConfiguration & configuration)
+{
+
+  m_thresholding = Thresholding();
```
There is no need to assign a new instance of `Thresholding`  to `m_thresholding`.
You can remove it from here.

joaobrittoneto commented on src/components/SignalGeneratorComponent.cpp:
```diff
@@ -0,0 +1,120 @@
+  hyro::Signal message_signal;
+  time_t t = std::time(0);
+
+  message_signal.timestamp = t;
```
You can do ` message_signal.timestamp = TimeUtils::Now_TimeSinceEpoch();`

Commit:
Removed unused variables and comments
Changed time variable from long int to hyro::Time
Changed cpp functions styles
Added break condition to exit program
Added code to test whether the value generated by DigitalConverterComponent is correct or not
Changed m_dummy variable to auto
